### PR TITLE
Add audit impl for rel=noopener for external links

### LINKED ISSUE #874: DBW: Suggestion rel="noopener" for external links
This one may be controversial. But links that open in a new tab should use `<a target="_blank" rel="noopener">` so the opening page's perf does not suffer.

https://jakearchibald.com/2016/performance-benefits-of-rel-noopener/

### PR DESCRIPTION
**Pull request by karanjthakkar**
Fixes #874.

cc: @ebidel

### CODE REVIEW

File: lighthouse-core/audits/dobetterweb/no-target-blank-links-without-rel-noopener.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+class NoTargetBlankLinksWithoutRelNoopenerAudit extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      category: 'Performance',
+      name: 'no-target-blank-links-without-rel-noopener',
+      description: 'Site does not have anchor links with target="_blank" and no rel="noopener"',
```
Comment (ebidel) [inline lighthouse-core/audits/dobetterweb/no-target-blank-links-without-rel-noopener.js]: Site opens external anchors using rel="noopener".

File: lighthouse-core/audits/dobetterweb/no-target-blank-links-without-rel-noopener.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      helpText: 'links that open in a new tab should use <code>target="_blank"</code> and <code>rel="noopener"</code> so the <a href="https://jakearchibald.com/2016/performance-benefits-of-rel-noopener">opening page\'s perf does not suffer.</a>',
```
Comment (ebidel) [inline lighthouse-core/audits/dobetterweb/no-target-blank-links-without-rel-noopener.js]: perf -> performance

File: lighthouse-core/audits/dobetterweb/no-target-blank-links-without-rel-noopener.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+   * @return {!AuditResult}
+   */
+  static audit(artifacts) {
+    const failingNodes =
+        artifacts.Links.filter(result => {
+          return result.target === '_blank' && result.rel !== 'noopener';
+        });
+
+    if (failingNodes.length > 0) {
+      return NoTargetBlankLinksWithoutRelNoopenerAudit.generateAuditResult({
```
Comment (ebidel) [inline lighthouse-core/audits/dobetterweb/no-target-blank-links-without-rel-noopener.js]: lets use the urlformatter to present the list of failing `<a>`. Example here: https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/audits/dobetterweb/no-datenow.js#L62-L74

File: lighthouse-core/audits/dobetterweb/no-target-blank-links-without-rel-noopener.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const Audit = require('../audit');
+
+class NoTargetBlankLinksWithoutRelNoopenerAudit extends Audit {
```
Comment (ebidel) [inline lighthouse-core/audits/dobetterweb/no-target-blank-links-without-rel-noopener.js]: suggest `ExternalAnchorsUseRelNoopenerAudit`

File: lighthouse-core/config/dobetterweb.json [hunk 5]
```diff
@@ -11,7 +11,8 @@
       "../gather/gatherers/dobetterweb/document-event-listeners",
       "../gather/gatherers/dobetterweb/document-write",
       "../gather/gatherers/dobetterweb/links-blocking-first-paint",
-      "../gather/gatherers/dobetterweb/websql"
+      "../gather/gatherers/dobetterweb/websql",
+      "../gather/gatherers/dobetterweb/links"
```
Comment (ebidel) [inline lighthouse-core/config/dobetterweb.json]: please keep alphabetical

File: lighthouse-core/config/dobetterweb.json [hunk 6]
```diff
@@ -26,6 +27,7 @@
     "../audits/dobetterweb/no-websql",
     "../audits/dobetterweb/uses-http2",
     "../audits/dobetterweb/uses-passive-event-listeners",
+    "../audits/dobetterweb/no-target-blank-links-without-rel-noopener",
```
Comment (ebidel) [inline lighthouse-core/config/dobetterweb.json]: please keep alphabetical

File: lighthouse-core/gather/gatherers/dobetterweb/links.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+'use strict';
+
+const Gatherer = require('../gatherer');
+
+function getPageLinks() {
+  return new Promise((resolve, reject) => {
+    const linkList = [...document.querySelectorAll('a')]
+        .map((link) => {
+          return {
+            target: link.getAttribute('target'),
```
Comment (ebidel) [inline lighthouse-core/gather/gatherers/dobetterweb/links.js]: target: link.target

File: lighthouse-core/gather/gatherers/dobetterweb/links.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            rel: link.getAttribute('rel')
```
Comment (ebidel) [inline lighthouse-core/gather/gatherers/dobetterweb/links.js]: rel: link.rel

File: lighthouse-core/gather/gatherers/dobetterweb/links.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    resolve(linkList);
+  });
+}
+
+class Links extends Gatherer {
+
+  afterPass(options) {
+    const driver = options.driver;
+    const scriptStr = `(${getPageLinks.toString()}())`;
+    return driver.evaluateAsync(scriptStr)
```
Comment (ebidel) [inline lighthouse-core/gather/gatherers/dobetterweb/links.js]: we really need to implement a qSA driver method like https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/gather/gatherers/viewport.js#L30. That would simplify this a bit.

File: lighthouse-core/gather/gatherers/dobetterweb/links.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+      .then(nodeList => {
+        this.artifact = nodeList;
+      })
+      .catch(_ => {
+        this.artifact = [];
```
Comment (ebidel) [inline lighthouse-core/gather/gatherers/dobetterweb/links.js]: `this.artifact = -1;` to signify an error condition

File: lighthouse-core/gather/gatherers/dobetterweb/links.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const Gatherer = require('../gatherer');
+
+function getPageLinks() {
+  return new Promise((resolve, reject) => {
+    const linkList = [...document.querySelectorAll('a')]
```
Comment (ebidel) [inline lighthouse-core/gather/gatherers/dobetterweb/links.js]: We can simplify the filtering a bit if you just get the relevant links directly from the DOM:

`document.querySelectorAll('a[target="_blank"]:not([rel="noopener"])')`

Comment (paulirish) [inline lighthouse-core/gather/gatherers/dobetterweb/links.js]: OTOH, I kinda like the idea of keeping the links gatherer general, in case another audit can make use of it.

File: lighthouse-core/gather/gatherers/dobetterweb/links.js (hunk 7 above)
Comment (paulirish) [inline lighthouse-core/gather/gatherers/dobetterweb/links.js]: agree. property could have been set by JS after HTML was parsed and takes priority over the attribute state. Good call.

**Comment by paulirish** [comment]
I noticed this PR doesn't consider if links are "external". I was curious if that mattered...

-----

Reading a bit about `noopener` (and testing a few things), it seems that it works on both same origin and external links. 

##### cross origin `target=_blank` link
* security: yeah you want it secure. noopener plz
* perf: yeah sg. 

##### same origin `target=_blank` link
* security: if you're using target=_blank you probably don't have security concerns, you just think a new tab is better UX. 
* perf: `noopener` will force the new tab into a separate process. Yes, even in the same domain. But only if `target=_blank` is also applied. There are a lot of optimizations to keeping multiple same origin tabs within the same process (shared caches, etc), **so I don't think it's worth it.** 
  * The only possible exception I can think of: you expect the user to have two pages open simultaneously and would rather use 2X of their RAM than allow the pages to potentially jank eachother. (and so on for three pages and 3X the RAM, etc)

Sidenote: `noreferrer` used to be [the unofficial way to do process forking for new windows](https://blog.chromium.org/2009/12/links-that-open-in-new-processes.html), but looks like `noopener` is the new blessed path.

(btw You can test process sharing by looking at the chrome task manager... )
image

-------

##### in conclusion 
Knowing all this now, I'd say that this PR should definitely check to see that the links are external before including them.

File: lighthouse-core/audits/dobetterweb/no-target-blank-links-without-rel-noopener.js (hunk 1 above)
Comment (karanjthakkar) [inline lighthouse-core/audits/dobetterweb/no-target-blank-links-without-rel-noopener.js]: done

File: lighthouse-core/audits/dobetterweb/no-target-blank-links-without-rel-noopener.js (hunk 2 above)
Comment (karanjthakkar) [inline lighthouse-core/audits/dobetterweb/no-target-blank-links-without-rel-noopener.js]: done

File: lighthouse-core/config/dobetterweb.json (hunk 5 above)
Comment (karanjthakkar) [inline lighthouse-core/config/dobetterweb.json]: done

File: lighthouse-core/gather/gatherers/dobetterweb/links.js (hunk 11 above)
Comment (karanjthakkar) [inline lighthouse-core/gather/gatherers/dobetterweb/links.js]: @ebidel I implemented your change and changed the file name to be more specific since this was residing in the `dobetterweb` folder. 

@paulirish I initially implemented this for the same reason but then the result returned by `evaluateAsync` kindda stripped down the `HTMLElement` object to an empty one. So I had to manually return the two attributes that I wanted to use. This kindda defeated the purpose of a common gatherer. So I did the changes as mentioned above.

File: lighthouse-core/gather/gatherers/dobetterweb/links.js (hunk 7 above)
Comment (karanjthakkar) [inline lighthouse-core/gather/gatherers/dobetterweb/links.js]: done

File: lighthouse-core/gather/gatherers/dobetterweb/links.js (hunk 8 above)
Comment (karanjthakkar) [inline lighthouse-core/gather/gatherers/dobetterweb/links.js]: done

File: lighthouse-core/audits/dobetterweb/no-target-blank-links-without-rel-noopener.js (hunk 3 above)
Comment (karanjthakkar) [inline lighthouse-core/audits/dobetterweb/no-target-blank-links-without-rel-noopener.js]: I've added the formatter to show the href's of the failing anchor links. I'm not sure what would be the correct thing to show here. Should we the entire anchor tag as a string? Or construct our own string?

File: lighthouse-core/gather/gatherers/dobetterweb/links.js (hunk 9 above)
Comment (karanjthakkar) [inline lighthouse-core/gather/gatherers/dobetterweb/links.js]: I tried writing the driver for that yesterday but couldn't get it together because there were some unknowns to me. I've written it now and moved the implementation to use that instead of `evaluateAsync`.

**Comment by karanjthakkar** [comment]
@paulirish I'm not sure just how much the scope of same origin is. Are anchors with href `xyz.domain.com` considered same origin for page on `domain.com`? also is vice versa true?

_OR_

Is the `anchorLink.hostname === document.location.hostname` check enough for same origin?

**Comment by ebidel** [comment]
@karanjthakkar we've been using [this check](https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/audits/dobetterweb/no-console-time.js#L56) in other places. You could use that to determine if the link is external.

File: lighthouse-core/audits/dobetterweb/external-anchors-use-rel-noopener.js [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+        artifacts.ExternalAnchorsWithNoRelNoopener === -1) {
+      return ExternalAnchorsUseRelNoopenerAudit.generateAuditResult({
+        rawValue: -1,
+        debugString: 'ExternalAnchorsWithNoRelNoopener gatherer did not run'
+      });
+    }
+
+    const failingAnchors = artifacts.ExternalAnchorsWithNoRelNoopener.usages.map(anchorHref => {
+      return {
+        url: anchorHref
```
Comment (ebidel) [inline lighthouse-core/audits/dobetterweb/external-anchors-use-rel-noopener.js]: This will just show anchors' URLs  when presented by the URL formatter. Not sure how helpful that will be. How about re-constructing the author's `<a>` snippet. If you had the actual `HTMLAnchorElement`s from the gatherer instead of just URLs strings, you could do:

url: \`${anchor.outerHTML}\`

File: lighthouse-core/gather/driver.js [hunk 13]
```diff
@@ -413,6 +413,28 @@ class Driver {
...
+        selector
+      }))
+      .then(nodeList => {
+        const elementList = [];
+        nodeList.nodeIds.forEach(nodeId => {
+          if (nodeId !== 0) {
+            elementList.push(new Element({nodeId}, this));
+          }
+        });
+        return Promise.all(elementList);
```
Comment (ebidel) [inline lighthouse-core/gather/driver.js]: `return elementList` would be sufficent, yea?

File: lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+const Gatherer = require('../gatherer');
+
+class ExternalAnchorsWithNoRelNoopener extends Gatherer {
+
+  afterPass(options) {
+    const driver = options.driver;
+    return driver.querySelectorAll('a[target="_blank"]:not([rel="noopener"])')
+      .then(failingNodeList => {
+        const failingNodes = failingNodeList.map(node => node.getAttribute('href'));
```
Comment (ebidel) [inline lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js]: `node.getAttribute('href')` -> `node.href` will get you the full resolved URL.

File: lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+        return Promise.all(failingNodes);
```
Comment (ebidel) [inline lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js]: Just return the val: `return failingNodes`

File: lighthouse-core/config/dobetterweb.json [hunk 16]
```diff
@@ -89,8 +91,13 @@
         "link-blocking-first-paint": {
           "expectedValue": false
         },
+<<<<<<< HEAD
         "external-anchors-use-rel-noopener": {
           "expectedValue": true
+=======
```
Comment (ebidel) [inline lighthouse-core/config/dobetterweb.json]: uh oh :)

File: lighthouse-core/gather/driver.js [hunk 17]
```diff
@@ -413,6 +413,28 @@ class Driver {
       });
   }
 
+  /**
+   * @param {string} selector Selector to find in the DOM
+   * @return {!Promise<Element[]>} The found elements, or [], resolved in a promise
+   */
+  querySelectorAll(selector) {
```
Comment (ebidel) [inline lighthouse-core/gather/driver.js]: nice job on the refactor. This will be useful elsewhere!

File: lighthouse-core/audits/dobetterweb/external-anchors-use-rel-noopener.js (hunk 12 above)
Comment (karanjthakkar) [inline lighthouse-core/audits/dobetterweb/external-anchors-use-rel-noopener.js]: @ebidel Yeah. I was thinking of using it, but if the authors anchor has a lot of nested DOM elements then the output on the results page might get too bloated. Thoughts?

File: lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js (hunk 14 above)
Comment (karanjthakkar) [inline lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js]: Are you sure? Because `node.href` is not defined on the `Element` class. 🤔

File: lighthouse-core/config/dobetterweb.json (hunk 16 above)
Comment (karanjthakkar) [inline lighthouse-core/config/dobetterweb.json]: fixed

File: lighthouse-core/audits/dobetterweb/external-anchors-use-rel-noopener.js (hunk 12 above)
Comment (ebidel) [inline lighthouse-core/audits/dobetterweb/external-anchors-use-rel-noopener.js]: How about reconstructing it then? 

url: \`<a href="${a.href}" ${a.target ? 'target="' + a.target + '"' : ''} ${a.rel ? ...}>`

File: lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js (hunk 14 above)
Comment (ebidel) [inline lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js]: I see. Normally you'd get at list of `HTMLAnchorElement` from `qSA('a')`, but I see the driver is using Element. `node.getAttribute('href')` it is then :)

File: lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js (hunk 15 above)
Comment (karanjthakkar) [inline lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js]: Since we're using `node.getAttribute` which returns a promise, this needs to be wrapped around `Promise.all`.

File: lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js (hunk 14 above)
Comment (karanjthakkar) [inline lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js]: 👍

File: lighthouse-core/audits/dobetterweb/external-anchors-use-rel-noopener.js (hunk 12 above)
Comment (karanjthakkar) [inline lighthouse-core/audits/dobetterweb/external-anchors-use-rel-noopener.js]: Sounds good. I'll add an implementation that will construct a string with those three attributes.

File: lighthouse-core/gather/driver.js (hunk 13 above)
Comment (ebidel) [inline lighthouse-core/gather/driver.js]: Right, but `elementList` is already the list of elements.

File: lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js (hunk 15 above)
Comment (ebidel) [inline lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js]: Forgot we were in driver code and ` node.getAttribute` is a promise here :) You're right. For this one you do need `Promise.all` or a `failingNodeList.reduce` or the promises.

File: lighthouse-core/gather/driver.js (hunk 13 above)
Comment (karanjthakkar) [inline lighthouse-core/gather/driver.js]: Ahh, you're right. Silly of me.

File: lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const Gatherer = require('../gatherer');
+
+class ExternalAnchorsWithNoRelNoopener extends Gatherer {
```
Comment (paulirish) [inline lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js]: Thx for adding the same-host check (technically just "same host" but not "same origin" as that [would include port like so](https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/audits/service-worker.js#L23-L31) -- i think it's okay to not consider port in this case. :) 

At this point the gatherer is just collecting ALL anchors with norelnoopener, so I think we should nuke the "External" in the name.

File: lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js (hunk 14 above)
Comment (paulirish) [inline lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js]: Im still a little suspect of this.  `url.parse('/thing/okay.html')` will deliver a host of `null`. 

I'd be happier if we did the work to get fully resolved URLs. I'll leave a larger comment below on how we can do that, if folks agree.

**Comment by paulirish** [comment]
#### getting node properties

If ya'll agree on getting the `node.href` property, then here's how we can do it via the protocol.

Here's what it looks like in the protocol when in the devtools' Element's panel you click on an element then click on the `Properties` pane:
image

Basically it looks like`DOM.resolveNode({nodeId: xxx})` will get us an objectId and then we use that on [`Runtime.getProperties`](https://chromedevtools.github.io/debugger-protocol-viewer/tot/Runtime/#method-getProperties) with the correct options and we're good:

image

image

**Comment by ebidel** [comment]
That looks mighty fine @paulirish. We should definitely get the resolved URL because a lot of authors will use relative paths in links.

File: lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js (hunk 18 above)
Comment (karanjthakkar) [inline lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js]: @paulirish The selector is `'a[target="_blank"]:not([rel="noopener"])'` so its actually gathering all external links with no `rel=noopener` attribute.

**Comment by karanjthakkar** [comment]
@paulirish @ebidel I'm not sure if adding a generic `getProperty` method on the `Element` class makes more sense or would a specific method for fetching the full href of an anchor tag be enough? Are there other cases where fetching runtime properties for elements would be necessary? In that case we can keep the generic implementation.

File: lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js (hunk 18 above)
Comment (karanjthakkar) [inline lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js]: Ahhh, you mean external as non same-host links? I added the `External` prefix thinking that its any link that opens in a new page.

**Comment by ebidel** [comment]
> adding a generic `getProperty` method on the `Element` class makes more sense

IMO, that would be useful. I can imagine this is something we'll need more of (getting JS properties off DOM nodes). Want to go this route? Unless @paulirish disagrees.

**Comment by ebidel** [comment]
"Good first bug" is turning out to be add awesome core features :)

**Comment by karanjthakkar** [comment]
@ebidel Well, I did go the generic route itself, if you'll check the implementation. Just wanted to know if thats what we wanted.

And yeah, didn't think it would extend to cover so many things 🙈

**Comment by paulirish** [comment]
> adding a generic `getProperty` method on the `Element` class makes more sense

Yes and let's add the _more_ generic getProperty on `Driver` as well. :) I bet we'll use that soon enough.

something like...

* driver.js's `getObjectProperty` will take an `objectId` and propName and return the value
* element.js's `getProperty` will take a `nodeId` and propName, leverage driver.getObjectProperty, and return the value

sound good?

---
@karanjthakkar, really excited about this PR. thanks for working with us despite all the back n forth! :)

**Comment by karanjthakkar** [comment]
@paulirish Yep. Done. 💯

File: lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js (hunk 18 above)
Comment (ebidel) [inline lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js]: Yes. @paulirish's bit [here](https://github.com/GoogleChrome/lighthouse/pull/912#issuecomment-258655127) explains the reasoning. If users are opening their own site's links  (e.g. same origin) using `target="_blank"` then that's fine. They're in control of their own perf, but this audit is really for external links that kill the opening page's perf.

Comment (karanjthakkar) [inline lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js]: Hmm. Makes sense. Guess I'll have to rename the gatherer file as well.

File: lighthouse-core/gather/driver.js [hunk 19]
```diff
@@ -394,6 +394,23 @@ class Driver {
       .then(_ => waitForLoad && this._waitForFullyLoaded(pauseAfterLoadMs));
   }
 
+  getObjectProperty(objectId, name) {
```
Comment (ebidel) [inline lighthouse-core/gather/driver.js]: can you add a jsdocs to this method?

File: lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js (hunk 18 above)
Comment (karanjthakkar) [inline lighthouse-core/gather/gatherers/dobetterweb/external-anchors-with-no-rel-noopener.js]: done

File: lighthouse-core/gather/driver.js [hunk 20]
```diff
@@ -395,6 +395,28 @@ class Driver {
...
+    return new Promise(resolve => {
+      this.sendCommand('Runtime.getProperties', {
+        objectId,
+        accessorPropertiesOnly: true,
+        generatePreview: false,
+        ownProperties: false,
+      })
+      .then(properties => {
+        const propertyForName = properties.result
+          .find(property => property.name === propName)
```
Comment (ebidel) [inline lighthouse-core/gather/driver.js]: think we need to be more careful here. `.find()` returns `undefined` if the property name can't be found.

Comment (karanjthakkar) [inline lighthouse-core/gather/driver.js]: Done. Does it make more sense now? :)

File: lighthouse-core/audits/dobetterweb/external-anchors-use-rel-noopener.js [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+      });
+    }
+
+    const pageHost = url.parse(artifacts.URL.finalUrl).host;
+    // Filter usages to exclude anchors that are same origin
+    const failingAnchors = artifacts.AnchorsWithNoRelNoopener.usages
+      .filter(anchorHref => url.parse(anchorHref).host !== pageHost)
+      .map(anchorHref => {
+        return {
+          url: anchorHref
```
Comment (ebidel) [inline lighthouse-core/audits/dobetterweb/external-anchors-use-rel-noopener.js]: Did you want to add https://github.com/GoogleChrome/lighthouse/pull/912#discussion_r86694878?

File: lighthouse-core/gather/driver.js (hunk 20 above)
Comment (ebidel) [inline lighthouse-core/gather/driver.js]: Great!

File: lighthouse-core/audits/dobetterweb/external-anchors-use-rel-noopener.js (hunk 21 above)
Comment (karanjthakkar) [inline lighthouse-core/audits/dobetterweb/external-anchors-use-rel-noopener.js]: Oh yes. I'll write it up once I finish the tests

Comment (karanjthakkar) [inline lighthouse-core/audits/dobetterweb/external-anchors-use-rel-noopener.js]: Done.

**Comment by karanjthakkar** [comment]
Phew. I thinnnk this should be all. I added tests and all the comments seem to be resolved. Gonna get some sleep now 😛 😴

**Comment by paulirish** [comment]
woohoo!

🎯 🐠 🐠 🐠 🐠 🐠 🐠

**Comment by karanjthakkar** [comment]
Thanks y'all! Glad to help 😸
